pennylane/templates/subroutines/qram.py

```diff
@@ -0,0 +1,421 @@
+            + list(portL_wires)
+            + list(portR_wires)
+        )
+
+        wire_manager = _QRAMWires(
+            qram_wires, target_wires, bus_wire, dir_wires, portL_wires, portR_wires
+        )
+
+        self._hyperparameters = {
+            "wire_manager": wire_manager,
+            "bitstrings": bitstrings,
+        }
+
+        super().__init__(wires=all_wires, id=id)
+
+    @classmethod
+    def _primitive_bind_call(cls, *args, **kwargs):
+        return cls._primitive.bind(*args, **kwargs)
+
+
+def _bucket_brigade_qram_resources(bitstrings):
+    num_target_wires = len(bitstrings[0])
+    num_qram_wires = int(math.log2(len(bitstrings)))
+    n_k = num_qram_wires
+    resources = defaultdict(int)
+    resources[resource_rep(SWAP)] = (
+        sum([1 if k == 0 else (1 << k) for k in range(n_k)]) + n_k
+    ) * 2 + num_target_wires * 2
+    resources[resource_rep(CSWAP)] = sum(
+        [(1 << ell) for ell in range(num_qram_wires)]
+    ) * num_target_wires * 2 + (sum([(1 << ell) for k in range(n_k) for ell in range(k)]) * 2)
+    resources[
+        controlled_resource_rep(
+            base_class=SWAP, base_params={}, num_control_wires=1, num_zero_control_values=1
+        )
+    ] = sum([(1 << ell) for ell in range(num_qram_wires)]) * num_target_wires * 2 + (
+        sum([(1 << ell) for k in range(n_k) for ell in range(k)]) * 2
+    )
+    resources[resource_rep(Hadamard)] += num_target_wires * 2
+    for j in range(num_target_wires):
+        for p in range(1 << n_k):
+            resources[resource_rep(PauliZ)] += 1 if int(bitstrings[p][j]) else 0
+    return resources
+
+
+def _mark_routers_via_bus_qfunc(wire_manager, n_k):
+    """Write low-order address bits into router directions **layer-by-layer** via the bus.
+
+    For each low bit a_k (k = 0..n_k-1):
+      1) SWAP(qram_wires[k], bus)
+      2) Route bus down k levels (CSWAPs controlled by routers at levels < k)
+      3) At node (k, path-prefix), SWAP(bus, dir[k, path-prefix])
+    """
+    for k in range(n_k):
+        # 1) load a_k into the bus
+        origin = wire_manager.qram_wires[k]
+        target = wire_manager.bus_wire[0]
+        SWAP(wires=[origin, target])
+        # 2) route down k levels
+        _route_bus_down_first_k_levels_qfunc(wire_manager, k)
+        # 3) deposit at level-k node on the active path
+        if k == 0:
+            SWAP(wires=[wire_manager.bus_wire[0], wire_manager.router(0, 0)])
+        else:
+            for p in range(1 << k):
+                # change to  in_wire later
+                parent = _node_index(k - 1, p >> 1)
+                if p % 2 == 0:
+                    origin = wire_manager.portL_wires[parent]
+                    target = wire_manager.router(k, p)
+                    SWAP(wires=[origin, target])
+                else:
+                    origin = wire_manager.portR_wires[parent]
+                    target = wire_manager.router(k, p)
+                    SWAP(wires=[origin, target])
```

Comment:
```suggestion
                origin = wire_manager.portL_wires[parent] if p % 2 == 0 else wire_manager.portR_wires[parent]
                target = wire_manager.router(k, p)
                SWAP(wires=[origin, target])
```
The same pattern applies in a few other places as well I believe.